quickevent/app/quickevent/plugins/CardReader/src/services/qropunch.cpp

```diff
@@ -0,0 +1,198 @@
+		} else {
+			qfInfo() << "Opened file for logging: " << ss.logFileName();
+		}
+	} else {
+		m_ofs.close();
+	}
+}
+
+void QrOPunch::onNewConnection()
+{
+	while (m_tcpServer->hasPendingConnections())
+	{
+		QTcpSocket *socket = m_tcpServer->nextPendingConnection();
+
+		connect(socket, &QTcpSocket::readyRead, this, &QrOPunch::onReadyRead);
+		connect(socket, &QTcpSocket::disconnected, this, [this] {
+			QTcpSocket *socket = static_cast<QTcpSocket*>(sender());
+			m_tcpClients.erase(socket);
+			socket->deleteLater();
+		});
+		m_tcpClients.emplace(socket, new QByteArray());
+	}
+}
+
+void QrOPunch::onReadyRead()
+{
+	QTcpSocket *socket = static_cast<QTcpSocket*>(sender());
+	auto sendError = [socket](const char *error) {
+		socket->write("HTTP/1.1 ");
+		socket->write(error);
+		socket->write("\r\n\r\n");
+		socket->flush();
+		socket->waitForBytesWritten(3000);
+		socket->close();
+	};
+	auto it = m_tcpClients.find(socket);
+	QByteArray *buffer = it->second.get();
+	while (socket->bytesAvailable() > 0)
+	{
+		buffer->append(socket->readAll());
+
+		// Parse the first request line
+		int eol = buffer->indexOf("\r\n");
+		if (-1 == eol) {
+			continue;
+		}
+		const static std::regex rex("^([A-Z]+) ([^:]+://[^/]*|)([^ ]+)");
+		std::cmatch match;
+		if (!std::regex_search(buffer->constData(), buffer->constData() + eol, match, rex)) {
+			sendError("405 Method Not Allowed");
+			continue;
+		}
+		if (match[1].str() != "POST") {
+			sendError("405 Method Not Allowed");
+			continue;
+		}
+		if (match[3].str() != "/card") {
+			sendError("404 Not Found");
+			continue;
+		}
+
+		// Figure out the whole request length and wait for it
+		const std::string SIZE = "Content-Length: ";
+		int lbegin = buffer->indexOf(SIZE.data());
+		if (-1 == lbegin) {
+			if (-1 != buffer->indexOf("\r\n\r\n")) {
+				sendError("405 Method Not Allowed");
+			}
+			continue;
+		}
+		lbegin += SIZE.size();
+		int lend = buffer->indexOf("\r", lbegin);
+		if (-1 == lend) {
+			continue;
+		}
+		std::string clength(buffer->constData() + lbegin, buffer->constData() + lend);
+		int len = std::stoi(clength);
+		int dataIdx = buffer->indexOf("\r\n\r\n");
+		if (-1 == dataIdx) {
+			continue;
+		}
+		dataIdx += 4;
+		if (buffer->size() < dataIdx + len) {
+			continue;
+		}
+
+		// Process POST /card
+		QByteArray data{buffer->mid(dataIdx, len)};
+		if (m_ofs) {
+			m_ofs << QDateTime::currentDateTime().toString("\ndd.MM.yyyy hh:mm:ss\n").toUtf8().constData();
+			m_ofs.write(buffer->constData(), dataIdx + len);
+			m_ofs << std::endl;
+		}
+		QJsonParseError parseError{};
+		QJsonDocument jsonDoc = QJsonDocument::fromJson(data, &parseError);
+		if (jsonDoc.isNull()) {
+			socket->write("HTTP/1.1 400 Bad Request\r\n\r\n");
+			socket->write("JSON error at the offset ");
+			socket->write(std::to_string(parseError.offset).c_str());
+			socket->write(":\n");
+			socket->write(parseError.errorString().toUtf8());
+			socket->flush();
+			socket->waitForBytesWritten(3000);
```

Comment:
do not use blocking calls, you can use lambda function connected to `QTcpSocket::onBytesWritten`

Reply:
Done